client/windows/CMapOverview.cpp

```diff
@@ -0,0 +1,240 @@
+{
+	logGlobal->debug("Building widget drawTransparentRect");
+
+	auto rect = readRect(config["rect"]);
+	auto color = readColor(config["color"]);
+	if(!config["colorLine"].isNull())
+	{
+		auto colorLine = readColor(config["colorLine"]);
+		return std::make_shared<TransparentFilledRectangle>(rect, color, colorLine);
+	}
+	return std::make_shared<TransparentFilledRectangle>(rect, color);
+}
+
+std::shared_ptr<CPicture> CMapOverview::CMapOverviewWidget::buildDrawMinimap(const JsonNode & config) const
+{
+	logGlobal->debug("Building widget drawMinimap");
+
+	auto rect = readRect(config["rect"]);
+	auto id = config["id"].Integer();
+
+	if(!renderImage)
+		return nullptr;
+
+	ResourcePath res = ResourcePath(parent.resource.getName(), EResType::MAP);
+	std::unique_ptr<CMap> campaignMap = nullptr;
+	if(parent.tabType != ESelectionScreen::newGame)
+	{
+		CLoadFile lf(*CResourceHandler::get()->getResourceName(ResourcePath(parent.resource.getName(), EResType::SAVEGAME)), MINIMAL_SERIALIZATION_VERSION);
+		lf.checkMagicBytes(SAVEGAME_MAGIC);
+
+		std::unique_ptr<CMapHeader> mapHeader = std::make_unique<CMapHeader>();
+		StartInfo * startInfo;
+		lf >> *(mapHeader) >> startInfo;
+
+		if(startInfo->campState)
+			campaignMap = startInfo->campState->getMap(*startInfo->campState->currentScenario());
+		res = ResourcePath(startInfo->fileURI, EResType::MAP);
+	}
+
+	std::vector<std::shared_ptr<IImage>> images;
+	if(!campaignMap)
+		images = createMinimaps(res, Point(rect.w, rect.h));
+	else
+		images = createMinimaps(campaignMap, Point(rect.w, rect.h));
+
+
+	if(id >= images.size())
+		return nullptr;
+
+	return std::make_shared<CPicture>(images[id], Point(rect.x, rect.y));
+}
+
+std::shared_ptr<CTextBox> CMapOverview::CMapOverviewWidget::buildDrawPath(const JsonNode & config) const
+{
+	logGlobal->debug("Building widget drawPath");
+
+	auto rect = readRect(config["rect"]);
+	auto font = readFont(config["font"]);
+	auto alignment = readTextAlignment(config["alignment"]);
+	auto color = readColor(config["color"]);
+
+	return std::make_shared<CTextBox>(parent.fileName, rect, 0, font, alignment, color);
+}
+
+std::shared_ptr<CLabel> CMapOverview::CMapOverviewWidget::buildDrawString(const JsonNode & config) const
+{
+	logGlobal->debug("Building widget drawString");
+
+	auto font = readFont(config["font"]);
+	auto alignment = readTextAlignment(config["alignment"]);
+	auto color = readColor(config["color"]);
+	std::string text = "";
+	if("mapname" == config["text"].String())
+		text = parent.mapName;
+	if("date" == config["text"].String())
+	{
+		if(parent.date.empty())
+		{
+			std::time_t time = boost::filesystem::last_write_time(*CResourceHandler::get()->getResourceName(ResourcePath(parent.resource.getName(), EResType::MAP)));
+			text = vstd::getFormattedDateTime(time);
+		}
+		else
+			text = parent.date;
+	}
+	auto position = readPosition(config["position"]);
+	return std::make_shared<CLabel>(position.x, position.y, font, alignment, color, text);
+}
```

Comment:
This is not a way how it supposed to be. You shouldn't create new object type every time when you need to override one field. Here it's sufficient to build simple CLabel with special name:
```json
...
{
  "type": "label",
  "name": "mapLabel",
  "text": "%mapname%"
}
```

Then in constructor
```c++
build(config);

if(auto w = widget<CLabel>("mapLabel"))
{
   //do special logic for text parsing
}
```

Reply:
if I use `w->setText()` I get a warning, that vcmi doesn't find localisation. How to fix?

Reply:
Ah, found the solution text needs to be "".